torch/utils/_sympy/value_ranges.py

```diff
@@ -9,21 +10,76 @@
 
 log = logging.getLogger(__name__)
 
+__all__ = ['ValueRanges', 'ValueRangeAnalysis']
+
+SympyBoolean = sympy.logic.boolalg.Boolean
+
+# Like sympify, but supports less stuff, and also ensures that direct
+# sympy expressions don't have free variables
+def simple_sympify(e):
+    if isinstance(e, int):
+        return sympy.Integer(e)
+    elif isinstance(e, float):
+        # infinity is special; we use it to bracket integers as well
+        if math.isinf(e):
+            return sympy.oo if e > 0 else -sympy.oo
+        return sympy.Float(e)
+    elif isinstance(e, bool):
+        return sympy.true if e else sympy.false
+    elif isinstance(e, sympy.Expr):
+        # TODO: Eventually, we will want to do indexing calculations with
+        # respect to symbols, so we can generate a dynamic kernel which will
+        # use 32-bit indexing so long as the dynamic dim isn't too big.  To do
+        # that, we will need to be able to do ValueRanges
+        assert not e.free_symbols, f"free variables NYI: {e}"
+        # NaNs can occur when doing things like 0 * sympy.oo, but it is better
+        # if the operator notices this and takes care of it, because sometimes
+        # the NaN is inappropriate (for example, for ints, the [-oo, oo] range
+        # should go to zero when multiplied with [0, 0])
+        assert e != sympy.nan
+        return e
+    elif isinstance(e, BooleanAtom):
+        return e
+    else:
+        raise AssertionError(f"not simple sympy type {type(e)}")
+
 @dataclasses.dataclass(frozen=True)
 class ValueRanges:
-    lower: Union[sympy.Expr, sympy.Number, int, float, bool]
-    upper: Union[sympy.Expr, sympy.Number, int, float, bool]
+    # Although the type signature here suggests you can pass any
+    # sympy expression, in practice the analysis here only works
+    # with sympy expressions with no free variables
+    lower: Union[sympy.Expr, SympyBoolean]
+    upper: Union[sympy.Expr, SympyBoolean]
+
+    def __init__(self, lower, upper):
+        lower = simple_sympify(lower)
+        upper = simple_sympify(upper)
+        # We don't support point-ranges on floating point inf
+        assert lower != sympy.oo
+        assert upper != -sympy.oo
+        # TODO: when the bounds have free variables, this may be
+        # nontrivial to actually verify
+        if isinstance(lower, sympy.Expr):
+            assert lower <= upper
+        else:
+            assert not (lower is sympy.true and upper is sympy.false)
+        # Because this is a frozen class
+        object.__setattr__(self, 'lower', lower)
+        object.__setattr__(self, 'upper', upper)
 
     def __contains__(self, x):
-        # TODO This needs to be generalised if lower/upper are sympy.Expr
-        assert not isinstance(x, sympy.Expr)
+        x = simple_sympify(x)
         return self.lower <= x <= self.upper
```

Comment:
i think we should still assert that the output is a bool. wydt ? I think user side people are going to not expect that this might return a value which cannot be used as a boolean. 

```
>>> import sympy
>>> from sympy import *
>>> x = Symbol("x")
>>> y = Symbol("y")
>>> bool(x <= y)
TypeError: cannot determine truth value of Relational
```